{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import glob\n",
    "import sklearn\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import re\n",
    "import seaborn as sns\n",
    "\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.model_selection import train_test_split\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.metrics import classification_report, roc_auc_score, roc_curve, cohen_kappa_score, precision_score, recall_score, accuracy_score\n",
    "from joblib import dump, load\n",
    "from sklearn.feature_selection import SelectFromModel, SelectPercentile, VarianceThreshold, chi2, f_classif\n",
    "\n",
    "plt.rc('figure', figsize=(16, 9))\n",
    "\n",
    "from sklearn.linear_model import SGDClassifier, LogisticRegression\n",
    "from sklearn.pipeline import make_pipeline, Pipeline\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.svm import LinearSVC\n",
    "from sklearn.preprocessing import MinMaxScaler, StandardScaler\n",
    "from sklearn.ensemble import ExtraTreesClassifier\n",
    "from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score, roc_curve, cohen_kappa_score, precision_score, recall_score, accuracy_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in raw files\n",
    "filenames=glob.glob(\"SNP_data_raw/5_95_data/final_data/*.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gffs_list=[]\n",
    "gffs_list.append('../Bioinfor2/gff/azt_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/cip_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/lev_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/cefo_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/mer_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/amp_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/tob_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/pip_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/cefe_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/tet_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/cefu_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/gen_ref.gff')\n",
    "gffs_list.append('../Bioinfor2/gff/tri_ref.gff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gff_to_df(gff_file):\n",
    "    with open(gff_file) as f:\n",
    "        content = f.readlines()\n",
    "\n",
    "    content = [x.strip() for x in content]\n",
    "    content_new=[]\n",
    "\n",
    "    for i in content:\n",
    "        if i=='##FASTA':\n",
    "            break\n",
    "        content_new.append(i)\n",
    "\n",
    "    content_list=[x.rstrip().split('\\t') for x in content_new[2:]]    \n",
    "\n",
    "    df=pd.DataFrame(content_list) \n",
    "    df.columns = ['a', 'b','c','d','e','f','g','h','i']\n",
    "\n",
    "    df=df[df.c=='gene']\n",
    "    a_list=[]\n",
    "    for x in df.i:\n",
    "        x=x.split(';')\n",
    "        if(len(x)==4):\n",
    "            a_list.append(x[2].split('=')[1])\n",
    "        else:\n",
    "            a_list.append(np.array([0])\n",
    "\n",
    "    df['gene']=a_list   \n",
    "    df.d=df.d.astype('int64')\n",
    "    df.e=df.e.astype('int64')\n",
    "\n",
    "    df=df[['d','e','gene']].rename(columns={\"d\": \"start\", \"e\": \"end\"})\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def top_weights_colormap(sim_all_df_T, top_num, X, y):\n",
    "\n",
    "    sim_all_df_Prod = sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num]\n",
    "    sim_all_df_Prod.drop(\"svm_weight_sum\", axis=1, inplace=True)\n",
    "\n",
    "    allele_to_colormap = []\n",
    "\n",
    "    clust_alleles = list(sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num].index)\n",
    "\n",
    "    allele_df = X.copy()[clust_alleles]\n",
    "    suscept_y = y[y==0].copy()\n",
    "    resist_y = y[y==1].copy()\n",
    "    suscept_alleles = allele_df.loc[suscept_y.index, :].copy()\n",
    "    resist_alleles = allele_df.loc[resist_y.index, :].copy()\n",
    "\n",
    "\n",
    "    allele_to_colormap = []\n",
    "    for allele in resist_alleles.columns:\n",
    "        allele_r_percent = round(resist_alleles[allele].sum()/float(allele_df[allele].sum()),2)\n",
    "        allele_to_colormap.append(sns.diverging_palette(250, 10, n=100)[int(allele_r_percent*100-1)])\n",
    "\n",
    "    return sim_all_df_Prod, allele_to_colormap, clust_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_heatmap(tmp_db, fgsz, allele_to_colormap, cmap_choose):\n",
    "    g = sns.clustermap(tmp_db, method = \"complete\", metric=\"cityblock\",\n",
    "                       yticklabels=True, xticklabels=True,\n",
    "                       cmap = cmap_choose,\n",
    "                       linewidths=.1,\n",
    "                       row_colors = allele_to_colormap,\n",
    "                       col_colors = allele_to_colormap,\n",
    "                       figsize=fgsz)\n",
    "    # figsize=(22,4)\n",
    "    plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "    plt.setp(g.ax_heatmap.xaxis.get_majorticklabels(), rotation=90);\n",
    "    return g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_gffs=dict(zip(filenames, gffs_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_gffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder='SVMs_paper3'\n",
    "import os\n",
    "if not os.path.exists(folder):\n",
    "    os.makedirs(folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open('anti_hyperparams_svm_ecoli.pickle', 'rb') as handle:\n",
    "    anti_hyperparams_svm = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# top_num: number of highly associated alleles to consider for downstream analysis 1:R\n",
    "top_num=50\n",
    "top_allele_weighted=200\n",
    "top_allele_pairs=100\n",
    "\n",
    "for k,v in file_gffs.items():\n",
    "    accuracy_list=[]\n",
    "    precision_score_list=[]\n",
    "    recall_score_list=[]\n",
    "    roc_auc_score_list=[]\n",
    "    cohenkappa_score_list=[]\n",
    "    tn_list=[]\n",
    "    fp_list=[]\n",
    "    fn_list=[]\n",
    "    tp_list=[]\n",
    "    specificity_list=[]\n",
    "    \n",
    "    antibiotic_name=k.split('/')[-1].split('_')[0]\n",
    "    \n",
    "    data=pd.read_csv(k)\n",
    "    X=data.iloc[:,1:-1]\n",
    "    y=data.iloc[:,-1]\n",
    "    sim_all_df = pd.DataFrame()\n",
    "\n",
    "    print('processing {}...'.format(antibiotic_name))\n",
    "    C_tunned=anti_hyperparams_svm[antibiotic_name]['clf__C']\n",
    "\n",
    "\n",
    "\n",
    "    for r in range(1,51):\n",
    "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=r)\n",
    "        \n",
    "    \n",
    "        model = Pipeline([\n",
    "          ('scaler', StandardScaler()),\n",
    "          ('clf', SVC(kernel='linear',C=C_tunned))\n",
    "        ])\n",
    "        \n",
    "        model.fit(X_train, y_train)\n",
    "        y_pred = model.predict(X_test)\n",
    "        predictions = [round(value) for value in y_pred]\n",
    "        \n",
    "#         sim_t_df = pd.DataFrame([model['clf'].coef_[0]], columns=X.columns[model['fs'].get_support()])\n",
    "        sim_t_df = pd.DataFrame([model['clf'].coef_[0]], columns=X.columns)\n",
    "\n",
    "        sim_all_df = pd.concat([sim_all_df, sim_t_df], ignore_index=True)\n",
    "        \n",
    "        accuracy_list.append(accuracy_score(y_test, predictions))\n",
    "        precision_score_list.append(precision_score(y_test,predictions))\n",
    "        recall_score_list.append(recall_score(y_test,predictions))\n",
    "        try:\n",
    "            roc_auc_score_list.append(roc_auc_score(y_test,predictions))\n",
    "        except ValueError:\n",
    "            roc_auc_score_list.append(-1)\n",
    "        cohenkappa_score_list.append(cohen_kappa_score(y_test,predictions))\n",
    "        tn, fp, fn, tp=confusion_matrix(y_test, predictions, labels=[0,1]).ravel()\n",
    "        tn_list.append(tn)\n",
    "        fp_list.append(fp)\n",
    "        fn_list.append(fn)\n",
    "        tp_list.append(tp)\n",
    "        specificity_list.append(tn / (tn+fp))\n",
    "\n",
    "\n",
    "    sim_all_df_T = sim_all_df.transpose().copy()\n",
    "    sim_all_df_T[\"svm_weight_sum\"] = sim_all_df_T.apply(lambda x: abs(x).sum(), axis=1)\n",
    "    sim_all_df_T_top = sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num]\n",
    "    sim_all_df_T_top.to_csv(\"{}/{}_SVMs_importances.csv\".format(folder,antibiotic_name))\n",
    "\n",
    "    sim_all_df_T_top.drop(\"svm_weight_sum\", axis=1, inplace=True)\n",
    "    sim_corr_df = sim_all_df_T_top.fillna(0).T.corr().copy()    \n",
    "    sim_corr_df = sim_corr_df[abs(sim_corr_df)>.05].fillna(0)\n",
    "    sim_corr_df.to_csv(\"{}/{}_correlation_matrix.csv\".format(folder,antibiotic_name))\n",
    "    \n",
    "    top_weighted_alleles = list(sim_all_df_T_top.index[:top_allele_weighted])\n",
    "    sim_corr_df = sim_corr_df.loc[top_weighted_alleles, top_weighted_alleles].copy()\n",
    "\n",
    "    \n",
    "    abs_corr_df = sim_corr_df.abs()\n",
    "    os = (abs_corr_df.where(np.triu(np.ones(abs_corr_df.shape), k=1).astype(np.bool))\n",
    "             .stack()\n",
    "             .sort_values(ascending=False))\n",
    "    \n",
    "    sig_allele_interact_list = []\n",
    "    for (allele_1, allele_2) in os.index[:top_allele_pairs]:\n",
    "        if allele_1 in top_weighted_alleles or allele_2 in top_weighted_alleles:\n",
    "                                    sig_allele_interact_list.append((allele_1, allele_2))\n",
    "    pd.DataFrame(sig_allele_interact_list).to_csv(\n",
    "        \"{}/{}_sig_allele_interact_list.csv\".format(folder,antibiotic_name))\n",
    "    \n",
    "    df=gff_to_df(v)\n",
    "    sig_gene_interact_list=[]\n",
    "    for (allele_1,allele_2) in sig_allele_interact_list:\n",
    "        sig_gene_interact_list.append(\n",
    "            (df[(df.start<int(allele_1)) & (df.end>int(allele_1))].gene.values,\n",
    "            df[(df.start<int(allele_2)) & (df.end>int(allele_2))].gene.values)\n",
    "        )\n",
    "    pd.DataFrame(sig_gene_interact_list).to_csv(\"{}/{}_sig_gene_interact_list.csv\".format(folder, antibiotic_name))\n",
    "    \n",
    "\n",
    "    stats = pd.DataFrame({'accuracy': accuracy_list, \n",
    "                              'precision': precision_score_list,\n",
    "                              'specificity':specificity_list,\n",
    "                              'sensitivity': recall_score_list,\n",
    "                              'auc': roc_auc_score_list,\n",
    "                              'kappa': cohenkappa_score_list,\n",
    "                                 'tn':tn_list,\n",
    "                                 'fp':fp_list,\n",
    "                                 'fn':fn_list,\n",
    "                                 'tp':tp_list\n",
    "                         })\n",
    "\n",
    "    stats.to_csv('{}/{}_SVM_stats.csv'.format(folder,antibiotic_name))\n",
    "    \n",
    "    sim_all_df_Prod, allele_to_colormap, clust_alleles=top_weights_colormap(sim_all_df_T, top_num, X, y)\n",
    "    sim_corr_df = sim_all_df_Prod.fillna(0).T.corr().copy()\n",
    "    sim_corr_df = sim_corr_df[abs(sim_corr_df)>.05].fillna(0)\n",
    "    gene_of_snp=[]\n",
    "    for snp in sim_corr_df.index:\n",
    "        gene=df[(df.start<int(snp)) & (df.end>int(snp))].gene.values\n",
    "        if gene!=0:\n",
    "            gene_of_snp.append(gene[0])\n",
    "        else:\n",
    "            gene_of_snp.append(snp)\n",
    "    sim_corr_df.index=gene_of_snp\n",
    "    sim_corr_df.columns=gene_of_snp\n",
    "    g_plot = plot_heatmap(sim_corr_df, (15, 15), allele_to_colormap, \"seismic\")\n",
    "    g_plot.savefig('{}/{}_correlation.png'.format(folder,antibiotic_name))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_all_df_T_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.clustermap(sim_all_df_Prod.fillna(0), \n",
    "                   col_cluster=False, row_cluster=True, cmap=\"seismic\",\n",
    "                  figsize=(15,15), xticklabels=False)\n",
    "plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "# g.savefig(SUPP_SVM_PLANES+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_iterations.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
